### PR TITLE
[dashboard] Refactor API using SIP-35

### PR DESCRIPTION
**Pull request by dpgaspar**
### CATEGORY

- [ ] Bug Fix
- [ ] Enhancement (new features, refinement)
- [X] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Refactor of the Dashboard API using SIP-35

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS
@willbarrett, @craig-rueda, @villebro

### CODE REVIEW

File: superset/dao/base.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+from superset.commands.exceptions import (
+    CreateFailedError,
+    DeleteFailedError,
+    OwnersNotFoundValidationError,
+    UpdateFailedError,
+)
+from superset.extensions import db
+
+
+def generic_create(model_cls: Model, properties: Dict, commit=True) -> Optional[Model]:
```
Comment (craig-rueda) [inline superset/dao/base.py]: Don't pass in the class here. Instead, the caller will have created whatever instance he wants and then should pass it in. You can then combine your create / update into a single `upsert` method. Also, you can call this something like `def upsert(model, commit=True)`

File: superset/dao/base.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        db.session.delete(model)
+        if commit:
+            db.session.commit()
+    except SQLAlchemyError as e:  # pragma: no cover
+        db.session.rollback()
+        raise DeleteFailedError(exception=e)
+    return model
+
+
+def get_owner_by_id(owner_id: int) -> Optional[object]:
```
Comment (craig-rueda) [inline superset/dao/base.py]: Not sure this belongs here. Instead, just define it on the `UserDao` as `def find_user_by_id(id: int)`

File: superset/dao/base.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    :return: User model or None
+    """
+    return (
+        db.session.query(current_app.appbuilder.sm.user_model)
+        .filter_by(id=owner_id)
+        .one_or_none()
+    )
+
+
+def populate_owners(user: User, owners_ids: Optional[List[int]] = None) -> List[User]:
```
Comment (craig-rueda) [inline superset/dao/base.py]: This is kinda business logic-y. Should be in a helper or something like that.

**Review by craig-rueda — CHANGES_REQUESTED** [review]

File: superset/dao/base.py (hunk 1 above)
Comment (dpgaspar) [reply superset/dao/base.py]: We can do that and end up with less code. Yet I tend to think it's more readable this way (and easier to reason), DAO create methods call `generic_create`, update methods call `generic_update`, also Exceptions follow the same pattern.

File: tests/dashboard_api_tests.py [hunk 4]
```diff
@@ -20,10 +20,11 @@
 from typing import List, Optional
```
Comment (willbarrett) [inline tests/dashboard_api_tests.py]: Would it be useful to move this file to `tests/dashboards/api_tests.py`? Following the implementation structure in the tests could help us find things in the future.

**Review by willbarrett — APPROVED** [review]
Responding to an existing discussion and one suggestion for module location, but overall this looks good!

**Review by craig-rueda — APPROVED** [review]

**Pull request merged**